reference/reference_test.go

```diff
@@ -0,0 +1,89 @@
+package reference
+
+import (
+	"strconv"
+	"testing"
+)
+
+func TestReferenceParse(t *testing.T) {
+	// referenceTestcases is a unified set of testcases for
+	// testing the parsing of references
+	referenceTestcases := []struct {
+		// input is the repository name or name component testcase
+		input string
+		// err is the error expected from ValidateRepositoryName, or nil
+		err error
+		// repository is the string representation for the reference
+		repository string
+		// hostname is the hostname expected in the reference
+		hostname string
+		// tag is the tag for the reference
+		tag string
+		// digest is the digest for the reference (enforces digest reference)
+		digest string
+	}{
```

Comment:
Added some more variations to add overlap with cases covered in the regexp test. Also added error throwing cases and checks on all fields.